src/electron/process_manager.ts

```diff
@@ -334,53 +359,125 @@ class ChildProcessHelper {
     if (!this.process) {
       // Never started.
       if (this.exitListener) {
-        this.exitListener();
+        this.exitListener(null);
       }
       return;
     }
 
     this.process.kill();
   }
 
-  set onExit(newListener: (() => void)|undefined) {
+  set onExit(newListener: ((code: number) => void)|undefined) {
     this.exitListener = newListener;
   }
 
   /**
    * Enables verbose logging for the process.  Must be called before launch().
    */
-  public enableDebugMode() {
+  enableDebugMode() {
     this.isInDebugMode = true;
   }
 }
 
-class SsLocal extends ChildProcessHelper {
-  constructor(private readonly proxyPort: number) {
+// shadowsocks-libev client program. Starts a local SOCKS proxy that relays TCP/UDP
+// traffic to a Shadowsocks proxy server.
+class SsLocal extends ChildProcessHelper implements ShadowsocksProcess {
+  // Construct with the Shadowsocks proxy config and the local port to bind the SOCKS proxy.
+  constructor(readonly config: ShadowsocksConfig, private readonly proxyPort: number) {
     super(pathToEmbeddedBinary('shadowsocks-libev', 'ss-local'));
   }
 
-  start(config: ShadowsocksConfig) {
-    // ss-local -s x.x.x.x -p 65336 -k mypassword -m aes-128-cfb -l 1081 -u
+  async start() {
+    // ss-local -s x.x.x.x -p 65336 -k mypassword -m chacha20-ietf-poly1035 -l 1081 -u
     const args = ['-l', this.proxyPort.toString()];
-    args.push('-s', config.host || '');
-    args.push('-p', '' + config.port);
-    args.push('-k', config.password || '');
-    args.push('-m', config.method || '');
+    args.push('-s', this.config.host || '');
+    args.push('-p', '' + this.config.port);
+    args.push('-k', this.config.password || '');
+    args.push('-m', this.config.method || '');
     args.push('-u');
     if (this.isInDebugMode) {
       args.push('-v');
     }
 
     this.launch(args);
+    return this.isSsLocalReachable();
+  }
+
+  async checkConnectivity() {
+    // Possibly over-cautious, though we have seen occasional failures immediately after network
+    // changes: ensure ss-local is reachable first.
+    await this.isSsLocalReachable();
+    // TODO(alalama): parallelize.
+    try {
+      await checkUdpForwardingEnabled(SSLOCAL_PROXY_ADDRESS, SSLOCAL_PROXY_PORT);
+    } catch (udpErr) {
+      await validateServerCredentials(SSLOCAL_PROXY_ADDRESS, SSLOCAL_PROXY_PORT);
+      throw udpErr;
+    }
+  }
+
+  private isSsLocalReachable() {
+    return isServerReachable(
+        SSLOCAL_PROXY_ADDRESS, SSLOCAL_PROXY_PORT, SSLOCAL_CONNECTION_TIMEOUT, SSLOCAL_MAX_ATTEMPTS,
+        SSLOCAL_RETRY_INTERVAL_MS);
   }
 }
 
-class Tun2socks extends ChildProcessHelper {
+// GoTun2socks implements the Shadowsocks protocol, so running a separate
+// Shadowsocks process is not necessary.
+// Stub process lifecycle to run connectivity checks through GoTun2socks.
+class GoShadowsocks implements ShadowsocksProcess {
```

Comment:
This abstraction is broken. GoShadowsocks is not a process.
We need to rethink this. Perhaps the combination of tun2socks + SS should all go in one, easily replaceable abstraction. You are currently exposing implementation details to the TunnelManager.